compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp

```diff
@@ -60,7 +61,23 @@ struct ConvertBatchMmt4DtoMmt4DPattern
     auto outType = out.getType().cast<RankedTensorType>();
     // Batch dim needs to be tiled to 1 first.
     if (outType.getShape()[0] != 1) {
-      return rewriter.notifyMatchFailure(op, "batch dim needs to be 1");
+      SmallVector<int64_t> tileSizes{1};
+      auto tilingInterfaceOp = cast<TilingInterface>(op.getOperation());
+      scf::SCFTileAndFuseOptions options{};
+      options.setTilingOptions(scf::SCFTilingOptions().setTileSizes(
+          getAsIndexOpFoldResult(op.getContext(), tileSizes)));
+      FailureOr<scf::SCFTileAndFuseResult> tiledResults =
+          scf::tileConsumerAndFuseProducerGreedilyUsingSCFForOp(
+              rewriter, tilingInterfaceOp, options);
+      if (failed(tiledResults))
+        return rewriter.notifyMatchFailure(op, "unable to tile the operation");
+      SmallVector<Value> replacements;
+      replacements.resize(op.getNumResults());
+      for (const auto &[index, result] : llvm::enumerate(op.getResults())) {
+        replacements[index] = tiledResults->replacements.lookup(result);
+      }
+      rewriter.replaceOp(op, replacements);
+      return success();
```

Comment:
I think it is better to decouple the logic from this pattern. We can either have a pattern to do the tiling or walkthrough batch_mmt4d ops (with funcOp.walk) and tile batch_dim with size=1. I prefer the latter one because we explicitly want to apply the pattern once. Can you move the logic to l.131 and do a func.walk? E.g.,

```
funcOp.walk([&](linalg::BatchMmt4DOp op) { ... } );
```

If we go with patterns, please create another pattern for this.

Reply:
Moved it to func.walk. Thanks :)